dpx/src/dpx_pdfdraw.rs

```diff
@@ -965,239 +793,176 @@ unsafe extern "C" fn pdf_dev__flushpath(
     pdf_doc_add_page_content(b, len as u32);
     0i32
 }
-unsafe extern "C" fn m_stack_init(mut stack: *mut m_stack) {
-    assert!(!stack.is_null());
-    (*stack).size = 0i32;
-    (*stack).top = 0 as *mut m_stack_elem;
-    (*stack).bottom = 0 as *mut m_stack_elem;
-}
-unsafe extern "C" fn m_stack_push(mut stack: *mut m_stack, mut data: *mut libc::c_void) {
-    let mut elem: *mut m_stack_elem = 0 as *mut m_stack_elem;
-    assert!(!stack.is_null());
-    elem = new((1_u64).wrapping_mul(::std::mem::size_of::<m_stack_elem>() as u64) as u32)
-        as *mut m_stack_elem;
-    (*elem).prev = (*stack).top;
-    (*elem).data = data;
-    (*stack).top = elem;
-    if (*stack).size == 0i32 {
-        (*stack).bottom = elem
-    }
-    (*stack).size += 1;
-}
-unsafe extern "C" fn m_stack_pop(mut stack: *mut m_stack) -> *mut libc::c_void {
-    let mut elem: *mut m_stack_elem = 0 as *mut m_stack_elem;
-    let mut data: *mut libc::c_void = 0 as *mut libc::c_void;
-    assert!(!stack.is_null());
-    if (*stack).size == 0i32 {
-        return 0 as *mut libc::c_void;
-    }
-    data = (*(*stack).top).data;
-    elem = (*stack).top;
-    (*stack).top = (*elem).prev;
-    if (*stack).size == 1i32 {
-        (*stack).bottom = 0 as *mut m_stack_elem
+
+trait Top<T> {
+    fn top(&mut self) -> &mut T;
+}
+
+impl<T> Top<T> for Vec<T> {
+    fn top(&mut self) -> &mut T {
+        let last = self.len() - 1;
+        &mut self[last]
     }
-    free(elem as *mut libc::c_void);
-    (*stack).size -= 1;
-    data
-}
-unsafe extern "C" fn m_stack_top(mut stack: *mut m_stack) -> *mut libc::c_void {
-    let mut data: *mut libc::c_void = 0 as *mut libc::c_void;
-    assert!(!stack.is_null());
-    if (*stack).size == 0i32 {
-        return 0 as *mut libc::c_void;
+}
+
+use lazy_static::lazy_static;
+use std::sync::Mutex;
+
+lazy_static! {
+    static ref gs_stack: Mutex<Vec<pdf_gstate>> = Mutex::new({
```

Comment:
i'm not very happy with the Mutex here. Can we add a struct called PdfDev or pdf_dev and collect these global states within it. and make the drawing invocations its method? (might be a big change to the code base)

Reply:
Helpful for seeing which statics are linked to by C++:

```sh
rg 'static mut (\w+):' myfile -o -r '$1' | xargs -n1 -I {} sh -c 'echo {} && rg -F {} c_sources_directory'
```

Generally this helps for moving definitions out of extern “C” when they don’t need to be in one.

Reply:
> i'm not very happy with the Mutex here. Can we add a struct called PdfDev or pdf_dev and collect these global states within it. and make the drawing invocations its method? (might be a big change to the code base)

I thought about this variant too. The problem as you say, it is a big change. And most likely we will be able to do this at one of the last steps of "oxidation".

Reply:
Mutex moved to start of file. Comment added.

Reply:
https://github.com/rust-lang/rust/pull/64028